src/main/java/minijava/token/Terminal.java

```diff
@@ -1,114 +1,92 @@
-  COMMA(",", SYNTAX_ELEMENT),
-  DOT(".", SYNTAX_ELEMENT),
-  RESERVED_OPERATORS("reserved operator", OPERATOR),
-  RESERVED_IDENTIFIER("reserved identifier", LITERAL),
-  BOOLEAN("boolean", TYPE),
-  INT("int", TYPE),
-  NEW("new", SYNTAX_ELEMENT),
-  RETURN("return", SYNTAX_ELEMENT),
-  THIS("this", LITERAL),
-  IF("if", CONTROL_FLOW),
-  WHILE("while", CONTROL_FLOW),
-  ELSE("else", CONTROL_FLOW),
-  TRUE("true", LITERAL),
-  FALSE("false", LITERAL),
-  PUBLIC("public", SYNTAX_ELEMENT),
-  STATIC("static", SYNTAX_ELEMENT),
-  VOID("void", TYPE),
-  NULL("null", LITERAL),
-  CLASS("class", SYNTAX_ELEMENT);
 
-  public static enum TerminalType {
-    OPERATOR,
-    TYPE,
-    LITERAL,
-    SYNTAX_ELEMENT,
-    CONTROL_FLOW,
-    HIDDEN,
-    MISC
-  }
+  // keywords
+  BOOLEAN("boolean"),
+  CLASS("class"),
+  ELSE("else"),
+  FALSE("false"),
+  IF("if"),
+  INT("int"),
+  NEW("new"),
+  NULL("null"),
+  PUBLIC("public"),
+  RETURN("return"),
+  STATIC("static"),
+  THIS("this"),
+  TRUE("true"),
+  VOID("void"),
+  WHILE("while"),
 
-  private final String description;
-  private final TerminalType terminalType;
-  private final int precedence;
-  private final boolean leftAssociative;
+  // operators
+  INVERT("!", LEFT, 0),
+  EQUAL_SIGN("=", RIGHT, 1),
+  OR("||", LEFT, 2),
+  AND("&&", LEFT, 3),
+  EQUALS("==", LEFT, 4),
+  UNEQUALS("!=", LEFT, 4),
+  LOWER("<", LEFT, 5),
+  LOWER_EQUALS("<=", LEFT, 5),
+  GREATER(">", LEFT, 5),
+  GREATER_EQUALS(">=", LEFT, 5),
+  PLUS("+", LEFT, 6),
+  MINUS("-", LEFT, 6),
+  MULTIPLY("*", LEFT, 7),
+  DIVIDE("/", LEFT, 7),
+  MODULO("%", LEFT, 7),
 
-  private Terminal(
-      String description, TerminalType terminalType, int precedence, boolean leftAssociative) {
-    this.description = description;
-    this.terminalType = terminalType;
-    this.precedence = precedence;
-    this.leftAssociative = leftAssociative;
-  }
+  // separators
+  LPAREN("("),
+  RPAREN(")"),
+  LBRACKET("["),
+  RBRACKET("]"),
+  LCURLY("{"),
+  RCURLY("}"),
+  DOT("."),
+  COMMA(","),
+  SEMICOLON(";"),
 
-  private Terminal(String description, TerminalType terminalType) {
-    this(description, terminalType, -1, true);
-  }
+  // with dynamic string values (lexval in Token is not null for tokens of this types)
+  IDENT,
+  INTEGER_LITERAL,
+  RESERVED, // reserved keyword or operator, parsing fails if tokens of this type exist
 
-  public String getDescription() {
-    return description;
-  }
+  // others - possible candidates for removal (except for EOF)!
+  EOF,
 
-  private static Terminal[] terminals = values();
+  COMMENT,
+  WS,
+  QUESTION_MARK("?"),
+  COLON(":");
 
-  public static Terminal valueOf(int id) {
-    return terminals[id];
+  public enum Associativity {
+    LEFT,
+    RIGHT
   }
 
-  public TerminalType getType() {
-    return terminalType;
-  }
+  public final Optional<String> string;
+  final Optional<Associativity> associativity;
+  final Optional<Integer> precedence;
```

Comment:
Well, the public field is OK I guess, but I found this rather interesting: http://blog.joda.org/2015/08/java-se-8-optional-pragmatic-approach.html

TLDR; Use `Optional` only as return value from public APIs, otherwise use `null` (and document so).